backend/src/apis/round.rs

```diff
@@ -60,60 +56,58 @@ impl MstInclusionProof {
 }
 
 pub struct Snapshot<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize> {
-    mst: MerkleSumTree<N_ASSETS, N_BYTES>,
+    mst: Box<dyn Tree<N_ASSETS, N_BYTES>>,
     assets_state: [Asset; N_ASSETS],
-    trusted_setup: [SetupArtifacts; 2],
+    trusted_setup: SetupArtifacts,
 }
 
-pub struct Round<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize> {
+pub struct Round<'a, const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize> {
     timestamp: u64,
     snapshot: Snapshot<LEVELS, N_ASSETS, N_BYTES>,
-    signer: SummaSigner,
+    signer: &'a SummaSigner,
 }
 
 impl<const LEVELS: usize, const N_ASSETS: usize, const N_BYTES: usize>
-    Round<LEVELS, N_ASSETS, N_BYTES>
+    Round<'_, LEVELS, N_ASSETS, N_BYTES>
 where
     [usize; N_ASSETS + 1]: Sized,
     [usize; 2 * (1 + N_ASSETS)]: Sized,
 {
-    pub fn new(
-        signer_key: &str,
-        chain_id: u64,
-        rpc_url: &str,
-        summa_address_input: AddressInput,
-        entry_csv_path: &str,
+    pub fn new<'a>(
+        signer: &'a SummaSigner,
+        mst: Box<dyn Tree<N_ASSETS, N_BYTES>>,
         asset_csv_path: &str,
         params_path: &str,
         timestamp: u64,
-    ) -> Result<Round<LEVELS, N_ASSETS, N_BYTES>, Box<dyn Error>> {
+    ) -> Result<Round<'a, LEVELS, N_ASSETS, N_BYTES>, Box<dyn Error>> {
         Ok(Round {
             timestamp,
-            snapshot: Snapshot::<LEVELS, N_ASSETS, N_BYTES>::new(
-                asset_csv_path,
-                entry_csv_path,
-                params_path,
-            )
-            .unwrap(),
-            signer: SummaSigner::new(signer_key, chain_id, rpc_url, summa_address_input),
+            snapshot: Snapshot::<LEVELS, N_ASSETS, N_BYTES>::new(mst, asset_csv_path, params_path)
+                .unwrap(),
+            signer: &signer,
         })
     }
 
     pub fn get_timestamp(&self) -> u64 {
         self.timestamp
     }
 
-    pub async fn dispatch_solvency_proof(&mut self) -> Result<(), Box<dyn Error>> {
-        let proof: SolvencyProof = match self.snapshot.generate_proof_of_solvency() {
-            Ok(p) => p,
-            Err(e) => return Err(format!("Failed to generate proof of solvency: {}", e).into()),
-        };
+    pub async fn dispatch_commitment(&mut self) -> Result<(), Box<dyn Error>> {
+        let root_str = format!("{:?}", self.snapshot.mst.root().hash);
+        let mst_root = U256::from_str_radix(&root_str, 16).unwrap();
+
+        let mut root_sums = Vec::<U256>::new();
+
+        for balance in self.snapshot.mst.root().balances.iter() {
+            let fp_str = format!("{:?}", balance);
+            root_sums.push(U256::from_str_radix(&fp_str, 16).unwrap());
+        }
 
         self.signer
-            .submit_proof_of_solvency(
-                proof.public_inputs[0],
+            .submit_commitment(
+                mst_root,
+                root_sums,
                 self.snapshot.assets_state.to_vec(),
```

Comment:
Something to think about for the future: assets come from the same source as the user balances. E.g., when the CEX is building the tree, it knows that it's now querying the ETH balances or BTC balances. That's why I don't think that the assets would live in a separate CSV file. Imagine that when the CEX is querying the balances, it specifies the assets in the query. If we're sticking to CSV, I would imagine a CSV that looks like `userId,balance_ETH_ETH,balance_USDT_ETH,...`. Won't request changes now but Interesting to know what you think @enricobottazzi @sifnoc

Reply:
It's not really an asset "state" per se, it's kind of a mapping of what assets are submitted in the MST to leave the record inside the transaction.

Reply:
I agree. It used to be a state when it also contained the asset_balances. Now it doesn't make sense for `asset.csv` to live in a separate file. Opening an issue.